### PR TITLE
Add RPC client

### PR DESCRIPTION
**Pull request by dnkolegov**
This PR adds a JSON RPC client to request IPC Agent, adds support for onchain reconfiguration, adds a basic test to check that onchain reconfiguration can be received via a stub JSOn RPC client, adapts CLI.

### CODE REVIEW

File: chain/consensus/mir/rpc/rpc.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+func NewJSONRPCClientWithContext(ctx context.Context, url, token string) *JSONRPCClient {
+	return &JSONRPCClient{
+		ctx:   ctx,
+		token: token,
+		url:   url,
+	}
+}
+
+func (c *JSONRPCClient) SendRequest(method string, params interface{}, reply interface{}) error {
+	paramBytes, err := rpc.EncodeClientRequest(method, params)
```
Comment (adlrocha) [inline chain/consensus/mir/rpc/rpc.go:50]: Are we sure that the method an parameters generated from this package are compatible with our implementation in the IPC agent?

File: chain/consensus/mir/rpc/rpc.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
+package rpc
```
Comment (adlrocha) [inline chain/consensus/mir/rpc/rpc.go:1]: Should we maybe move this `rcp` package to `chain/ipc-agent`? We may want to move it out to its own repo eventually (it is not lotus-specific) and is probably better if we don't keep it as part of mir here.

File: chain/consensus/mir/rpc/rpc.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	// Return an error for any nonsuccessful status code
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		// Drop any error during close to report the original error
+		_ = resp.Body.Close()
+		return fmt.Errorf("received status code: %d", resp.StatusCode)
+	}
+
+	if err := rpc.DecodeClientResponse(resp.Body, reply); err != nil {
```
Comment (adlrocha) [inline chain/consensus/mir/rpc/rpc.go:81]: Same here, we should ensure that is compatible with our implementation.

File: chain/consensus/mir/validator/membership.go [hunk 4]
```diff
@@ -32,3 +36,27 @@ type EnvMembership string
 func (e EnvMembership) GetValidatorSet() (*Set, error) {
 	return NewValidatorSetFromEnv(string(e))
 }
+
+// -----
+
+const JSONRPCServerURL = "http://127.0.0.1:3030"
```
Comment (adlrocha) [inline chain/consensus/mir/validator/membership.go:42]: We should figure out a way to pass this as a config parameter when we start the daemon. We should allow a way to pass the endpoint for the ipc-agent related to the lotus instance.

File: chain/consensus/mir/validator/membership.go [hunk 5]
```diff
@@ -32,3 +36,27 @@ type EnvMembership string
...
+func NewActorMembershipClient(client rpc.JSONRPCRequestSender) *ActorMembership {
+	return &ActorMembership{
+		client: client,
+	}
+}
+
+// GetValidatorSet gets the membership config from the actor state.
+func (c *ActorMembership) GetValidatorSet() (*Set, error) {
+	var set Set
+	err := c.client.SendRequest("Filecoin.GetValidatorSet", nil, &set)
```
Comment (adlrocha) [inline chain/consensus/mir/validator/membership.go:57]: This is wrong, the method name is the following: https://github.com/consensus-shipyard/ipc-agent/blob/5815f5955c1eaaf4bf07637e460cfbcee1f2bdf4/src/config/server.rs#L20

File: cmd/eudico/mirvalidator/run.go [hunk 6]
```diff
@@ -60,6 +61,16 @@ var runCmd = &cli.Command{
 			Name:  "init-checkpoint",
 			Usage: "pass initial checkpoint as a file (it overwrites 'init-height' flag)",
 		},
+		&cli.StringFlag{
```
Comment (adlrocha) [inline cmd/eudico/mirvalidator/run.go:64]: We should pass a flag to point to the ipc-agent endpoint and use this throughout the code when requiring an interaction with the agent. We should do the same also for the lotus daemon.

**Review by adlrocha — CHANGES_REQUESTED** [review]
Great work! A few suggestions and potential bugs. It would also be great if we can actually test it against the IPC agent to be sure that it works (at least the call, we can plan the end-to-end tests in a follow-up PR).

File: chain/consensus/mir/rpc/rpc.go (hunk 2 above)
Comment (dnkolegov) [reply chain/consensus/mir/rpc/rpc.go:1]: Do we have `chain/ipc-agent`, or do you want me to create that one?

File: chain/consensus/mir/rpc/rpc.go (hunk 1 above)
Comment (dnkolegov) [reply chain/consensus/mir/rpc/rpc.go:50]: Yes, I asked @cryptoAtwill to create a string-based test in the JSON RPC server repo. Then I can use the same strings (serialized requests) to test them here.
Do you have other ideas?

File: chain/consensus/mir/validator/membership.go (hunk 5 above)
Comment (dnkolegov) [reply chain/consensus/mir/validator/membership.go:57]: Fixed

File: cmd/eudico/mirvalidator/run.go (hunk 6 above)
Comment (dnkolegov) [reply cmd/eudico/mirvalidator/run.go:64]: Should we address that in a separate PR?

File: chain/consensus/mir/rpc/rpc.go (hunk 2 above)
Comment (adlrocha) [reply chain/consensus/mir/rpc/rpc.go:1]: We don't have it yet, can you create it?

File: chain/consensus/mir/rpc/rpc.go (hunk 1 above)
Comment (adlrocha) [reply chain/consensus/mir/rpc/rpc.go:50]: This works for now, just to double-check before we get integration tests ready.

File: cmd/eudico/mirvalidator/run.go (hunk 6 above)
Comment (adlrocha) [reply cmd/eudico/mirvalidator/run.go:64]: Sounds good, feel free to open an issue to track it and let's address it next. Thanks.

File: chain/consensus/mir/rpc/rpc.go (hunk 2 above)
Comment (dnkolegov) [reply chain/consensus/mir/rpc/rpc.go:1]: Sure

**Pull request merged**